src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificatePal.cs

```diff
@@ -19,82 +19,160 @@ public static ICertificatePal FromHandle(IntPtr handle)
             return new OpenSslX509CertificateReader(Interop.libcrypto.X509_dup(handle));
         }
 
-        public static unsafe ICertificatePal FromBlob(byte[] rawData, string password, X509KeyStorageFlags keyStorageFlags)
+        public static ICertificatePal FromBlob(byte[] rawData, string password, X509KeyStorageFlags keyStorageFlags)
         {
-            // If we can see a hyphen, assume it's PEM.  Otherwise try DER-X509, then fall back to DER-PKCS12.
-            SafeX509Handle cert;
+            ICertificatePal cert;
 
-            // PEM
-            if (rawData[0] == '-')
+            if (TryReadX509Der(rawData, out cert) ||
+                TryReadX509Pem(rawData, out cert) ||
+                PkcsFormatReader.TryReadPkcs7Der(rawData, out cert) ||
+                PkcsFormatReader.TryReadPkcs7Pem(rawData, out cert) ||
+                PkcsFormatReader.TryReadPkcs12(rawData, password, out cert))
             {
-                using (SafeBioHandle bio = Interop.libcrypto.BIO_new(Interop.libcrypto.BIO_s_mem()))
+                if (cert == null)
                 {
-                    Interop.libcrypto.CheckValidOpenSslHandle(bio);
-
-                    Interop.libcrypto.BIO_write(bio, rawData, rawData.Length);
-                    cert = Interop.libcrypto.PEM_read_bio_X509_AUX(bio, IntPtr.Zero, IntPtr.Zero, IntPtr.Zero);
+                    // Empty collection, most likely.
+                    throw new CryptographicException();
                 }
 
-                Interop.libcrypto.CheckValidOpenSslHandle(cert);
+                return cert;
+            }
+
+            // Unsupported
+            throw Interop.libcrypto.CreateOpenSslCryptographicException();
+        }
+
+        public static ICertificatePal FromFile(string fileName, string password, X509KeyStorageFlags keyStorageFlags)
+        {
+            // If we can't open the file, fail right away.
+            using (SafeBioHandle fileBio = Interop.libcrypto.BIO_new_file(fileName, "rb"))
+            {
+                Interop.libcrypto.CheckValidOpenSslHandle(fileBio);
+
+                return FromBio(fileBio, password);
+            }
+        }
+
+        private static ICertificatePal FromBio(SafeBioHandle bio, string password)
+        {
+            int bioPosition = Interop.Crypto.BioTell(bio);
+
+            Debug.Assert(bioPosition >= 0);
 
-                return new OpenSslX509CertificateReader(cert);
+            ICertificatePal certPal;
+            if (TryReadX509Pem(bio, out certPal))
+            {
+                return certPal;
             }
 
-            // DER-X509
-            cert = Interop.libcrypto.OpenSslD2I((ptr, b, i) => Interop.libcrypto.d2i_X509(ptr, b, i), rawData, checkHandle: false);
+            // Rewind, try again.
+            Interop.Crypto.BioSeek(bio, bioPosition);
```

Comment:
It does appear to have an ignored return value, which is negative on error.  Added a helper function for everything except the last (post-exception-creation-pre-expection-throw) rewind, which gained a comment.